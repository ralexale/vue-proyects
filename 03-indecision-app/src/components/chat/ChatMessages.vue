<template>
  <div ref="chatRef" class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <!-- Messages go here -->
      <ChatBubble data-testid="chat-bubble" v-for="message in $props.messages" :key="message.id" v-bind="message" />
    </div>
  </div>
</template>

<script setup lang="ts">
import type { ChatMessage } from '@/interfaces/chat-message.interface';
import ChatBubble from './ChatBubble.vue';
import { ref, watch } from 'vue';

interface Props {
  messages: ChatMessage[];
}

const props = defineProps<Props>();

const chatRef = ref<HTMLDivElement | null>(null);

watch(props, () => {
  setTimeout(() => {
    chatRef.value?.scrollTo({
      top: chatRef.value?.scrollHeight,
      behavior: 'smooth',
    });
  }, 100);
});
</script>
