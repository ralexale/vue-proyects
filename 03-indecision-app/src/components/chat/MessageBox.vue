<template>
  <div class="bg-white p-4 flex items-center">
    <input
      type="text"
      v-model="message"
      @keypress.enter="sendMessage"
      placeholder="Type your message..."
      class="flex-1 border rounded-full px-4 py-2 focus:outline-none"
    />
    <button
      @click="sendMessage"
      class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none"
    >
      <img src="/icons/chat/sendIcon.svg" alt="send icon" />
    </button>
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue';

const message = ref<string>('');

const emit = defineEmits<{
  sendMessage: [message: string];
}>();

const sendMessage = async () => {
  if (!message.value) return;
  emit('sendMessage', message.value);
  message.value = '';
};
</script>
