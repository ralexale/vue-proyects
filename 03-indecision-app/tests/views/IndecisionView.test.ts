import { mount } from '@vue/test-utils';
import ChatMessages from '@/components/chat/ChatMessages.vue';
import IndecisionView from '@/views/IndecisionView.vue';
import MessageBox from '@/components/chat/MessageBox.vue';

const mockChatMessages = {
  template: '<div data-testid="mock-messages">Mock ChatMessages</div>',
};
describe('<IndecisionView />', () => {
  test('should renders the chat messages and the message box', () => {
    const wrapper = mount(IndecisionView);
    const chatMessages = wrapper.findComponent(ChatMessages);
    const messageBox = wrapper.findComponent(MessageBox);

    expect(chatMessages.exists()).toBe(true);
    expect(messageBox.exists()).toBe(true);
  });

  test('calls on message when sending a message', async () => {
    const message = 'hola mundo';
    const wrapper = mount(IndecisionView, {
      global: {
        stubs: {
          ChatMessages: mockChatMessages,
        },
      },
    });
    // const chatMessages = wrapper.findComponent(ChatMessages);
    const messageBox = wrapper.findComponent(MessageBox);
    messageBox.vm.$emit('sendMessage', message);
    await new Promise((r) => setTimeout(r, 150));

    expect(messageBox.emitted('sendMessage')?.[0]).toEqual([message]);
    expect(wrapper.html()).toMatchSnapshot()
  });
});
