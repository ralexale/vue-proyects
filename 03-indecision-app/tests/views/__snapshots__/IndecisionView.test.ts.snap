// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<IndecisionView /> > calls on message when sending a message 1`] = `
"<div class="bg-gray-100 h-screen flex flex-col max-w-lg mx-auto">
  <div class="bg-blue-500 p-4 text-white flex justify-between items-center"><span>Persona</span></div>
  <div data-testid="mock-messages" messages="">Mock ChatMessages</div>
  <div class="bg-white p-4 flex items-center"><input type="text" placeholder="Type your message..." class="flex-1 border rounded-full px-4 py-2 focus:outline-none"><button class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none"><img src="/icons/chat/sendIcon.svg" alt="send icon"></button></div>
</div>"
`;
